adc/ad7779/ad7779-tool.c

```diff
@@ -0,0 +1,525 @@
+	volatile uint32_t *base = mmap(NULL, ioctl.buffers.size, PROT_READ, MAP_DEVICE, OID_PHYSMEM, ioctl.buffers.paddr);
+
+	if (base == MAP_FAILED) {
+		printf("failed to mmap buffers\n");
+		return 1;
+	}
+
+	for (unsigned i = 0; i < tool_common.config.nr; ++i) {
+		tool_common.config.ptr[i] = (uint32_t *)((uint8_t *)base + i * tool_common.config.size);
+	}
+
+	return 0;
+}
+
+
+static int adc_reset(uint32_t port)
+{
+
+	msg_t msg = { 0 };
+	adc_dev_ctl_t ioctl = { 0 };
+
+	msg.type = mtDevCtl;
+	ioctl.type = adc_dev_ctl__disable;
+	memcpy(msg.o.raw, &ioctl, sizeof(adc_dev_ctl_t));
+
+	log_verbose("disable\n");
+	if (msgSend(port, &msg) < 0 || msg.o.io.err != 0) {
+		printf("sample: failed to disable adc device: %d\n", msg.o.io.err);
+		return 1;
+	}
+
+	msg.type = mtDevCtl;
+	ioctl.type = adc_dev_ctl__reset;
+	ioctl.mux = 0;
+	memcpy(msg.o.raw, &ioctl, sizeof(adc_dev_ctl_t));
+
+	log_verbose("reset\n");
+	if (msgSend(port, &msg) < 0 || msg.o.io.err != 0) {
+		printf("sample: failed to reset adc device: %d\n", msg.o.io.err);
+		return 1;
+	}
+
+	msg.type = mtDevCtl;
+	ioctl.type = adc_dev_ctl__set_config;
+	ioctl.config.sampling_rate = 8000;
+	ioctl.config.enabled_ch = ACTIVE_CHANNELS;
+	memcpy(msg.o.raw, &ioctl, sizeof(adc_dev_ctl_t));
+
+	log_verbose("set_config\n");
+	if (msgSend(port, &msg) < 0 || msg.o.io.err != 0) {
+		printf("sample: failed to set sample rate: %d\n", msg.o.io.err);
+		return 1;
+	}
+
+	/* set channel gains */
+	for (int i = 0; i < PHASE_CNT * 2; i++) {
+		msg.type = mtDevCtl;
+		ioctl.type = adc_dev_ctl__set_channel_gain;
+
+		ioctl.gain.channel = i;
+
+		if (((1 << i) & ACTIVE_CHANNELS) == 0) {
+			continue;
+		}
+
+		ioctl.gain.val = adc_gain[i];
+		log_verbose("set_channel_gain ch=%u\n", i);
+		memcpy(msg.o.raw, &ioctl, sizeof(adc_dev_ctl_t));
+		if (msgSend(port, &msg) < 0 || msg.o.io.err != 0) {
+			printf("sample: failed to set channel %d gain: %d\n", i, msg.o.io.err);
+			return 1;
+		}
+	}
+
+	msg.type = mtDevCtl;
+	ioctl.type = adc_dev_ctl__enable;
+	memcpy(msg.o.raw, &ioctl, sizeof(adc_dev_ctl_t));
+
+	log_verbose("enable\n");
+	if (msgSend(port, &msg) < 0 || msg.o.io.err != 0) {
+		printf("sample: failed to enable adc: %d\n", msg.o.io.err);
+		return 1;
+	}
+
+	return 0;
+}
+
+
+static uint32_t read_intr(uint32_t port)
+{
+	msg_t msg = { 0 };
+	uint32_t intr = 0;
+
+	msg.type = mtRead;
+	msg.o.size = sizeof(uint32_t);
+	msg.o.data = &intr;
+
+	if (msgSend(port, &msg) || msg.o.io.err < 0) {
+		printf("MSG failed to read adc intr: %s", strerror(msg.o.io.err));
+	}
+
+	return intr;
+}
+
+
+static void signal_handler(int signal)
+{
+	tool_common.running = 0;
+}
+
+
+static inline uint8_t raw2channel(uint32_t raw)
+{
+	return (raw & 0x70000000) >> 28;
+}
+
+
+static inline int32_t raw2value(uint32_t raw)
+{
+	int32_t value = raw & 0x00FFFFFFu;
+	if (value >= 1 << 23)
+		value -= 1 << 24;
+	return value;
+}
+
+
+#define print_us(type_, us_) printf("time " type_ ": %llu.%03llu ms\n", (us_) / 1000, (us_) % 1000);
```

Comment:
```suggestion
#define PRINT_US(type_, us_) printf("time " type_ ": %llu.%03llu ms\n", (us_) / 1000, (us_) % 1000)
```

Reply:
switched to inline function, might be slightly less optimal but I don't particularly care in this case.